### PR TITLE
[DOC] Add tutorial for adding your own objects

### PR DESCRIPTION
**Pull request by smokestacklightnin**
This is a work in progress.

This is a continuation of PR #358. It was created by rebasing PR #358 to include the merged commits from PR #362 that removes `max_input_size`. 

This PR adds a tutorial to help users or developers add assistants and source storages for LLMs to Ragna that are not currently supported.

TODO: Add usage example with the Python API

### CODE REVIEW

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:6]: I would phrase this a little different. Ragna was built with the goal in mind that you bring your own components. We only provide very basic ones inside the library. For everything else you'll have to implement your own.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:2]: The title is visible in the side bar and thus should be as concise as possible.

```suggestion
# Custom Components
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:13]: 1. While probably true in practice, the `Assistant` doesn't have to be an LLM. We can put this as text before the actual implementation.
2. Let's keep the header concise.

```suggestion
# ## Assistant
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:24]: We can have a second look when we have the usage example and actually see the output, but I think we can get rid of this blank line.
```suggestion
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:39]: I would move these explanations before the implementation.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:38]: We should add a quick explanation what `Source`s actually are. Recall that this is likely the first time a user of Ragna actually comes in touch with them.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:34]:
```suggestion
# For example, this could call an API directly or call a local LLM.
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:48]: We should move this before the assistant, because

1. the source storage is used before the assistant in the RAG workflow
2. the source storage creates `Source`s that will be consumed by the assistant. If we explain what they are here, we can keep it short in the assistant example.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:56]:
```suggestion
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:20]: Before we actually put the code, we need to explain what the `TutorialAssistant` will do. Something along the lines of "for simplicity, we are not going to implement an actual LLM here, but rather a demo assistant that just mirrors back the inputs. This is similar as the RagnaDemoAssistant".

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:63]: Lose the trailing comma for the autoformatter to inline this

```suggestion
                document=document
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:60]: There are three things here:

1. The `chat_id` is not an `int`, but rather a `uuid.UUID`.
2. So far we have not explained what `chat_id` is and where it comes from. It is not part of the standard API:

    https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L102-L103

    https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L111-L112

    This deserves a full section below. Meaning at this point, we need to tell the reader "we need a way to identify the chat that we are storing documents for. We can use the unique `chat_id` for this. To get access to it, just add it as parameter to method signature. We'll explain how this works in section X".
3. It is convention to separate the non-standard parameters to be keyword-only.
    ```suggestion
        def store(self, documents: list[Document], *, chat_id: int) -> None:
    ```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:74]: Same as for the assistant. Move this before the implementation.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:87]: The usage example for the Python API should be added before this section.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:88]: We should rename this to reflect that you need to do this if you want to access your custom components from the REST API or web UI.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:109]: While correct, the regular way is for users to add the components to their configuration file, e.g. https://ragna.chat/en/stable/references/config/#referencing-python-objects.

**Review by pmeier — DISMISSED** [review]

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [reply docs/tutorials/gallery_adding_components.py:60]: Re 2.: Here is how it works:

When adding new parameters to the signature of the protocol methods

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L99-L100

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L210-L213

we automatically pick up on this and add the parameters to the model of the component

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_components.py#L73-L76

When creating a new `Chat`, we check if the additional parameters satisfy the additional parameters of the selected components:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L251

You don't need to understand this in detail. The mental model that you can use is that the extra `params` that you can pass to the `Chat`

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L77-L84

are automatically checked for completeness and distributed to the components at runtime. Meaning, every parameter you define without a default, must be passed when creating the chat. If you add a default value, this can be overridden by passing a value to the chat, but you don't have to.

So why didn't we need to pass the `chat_id` although it didn't have a default? `chat_id` is a special parameter alongside `chat_name` and `user`:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/core/_rag.py#L103-L108

They are automatically set by Ragna and can always be added to a signature without the need for the user to supply them. Same as the other parameters they can be overridden by passing other values when creating the chat. This is what we do for the REST API:

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/ragna/deploy/_api/core.py#L204-L207

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:166]: We don't need a full configuration file here. We can point to the reference in the documentation and just put the important configuration here. We don't want to drown the useful information in the noise.

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:228]:
```suggestion
config = Config(source_storages=[TutorialSourceStorage], assistants=[TutorialAssistant])
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:249]: This will give you a fully authenticated client

```suggestion
client = rest_api.start(config, authenticate=True)
```

File: docs/tutorials/gallery_adding_components.py
Comment (pmeier) [inline docs/tutorials/gallery_adding_components.py:254]: Since we have the disclaimer above on where to find detailed information, we can cut this short here. We don't need to explain in detail the parts in this tutorial that are necessary, but not interesting to drive the narrative.

```suggestion
# Next, we upload the documents.
```

File: docs/assets/images/ragna-tutorial-components.png
Comment (pmeier) [inline docs/assets/images/ragna-tutorial-components.png:1]: This needs to be a `git lfs` pointer rather than the actual image. See https://ragna.chat/en/stable/community/contribute/#view-add-or-update-images. I don't have push access to your fork, so I wasn't able to resolve this.

**Review by pmeier — COMMENTED** [review]
I've fixed most of my own comments. There is some slight cleanup needed, but this is good for a first round of review from someone else.

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:226]: I think we should clearly state what the _regular way_ is (i.e. say that instead of regular), since it's unclear to a reader here (e.g. me!).

Do you mean through the CLI's `api` command here? If so, let's state that here.

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:250]:
```suggestion
# parameters are defined by adding them to the signature of the method. For
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:253]:
```suggestion
# example, let's define a more elaborate assistant that accepts the following arguments:
#
# * `my_required_parameter` must be passed and has to be an `int`
# * `my_optional_parameter` may be passed and has to be a `str` if passed
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:274]:
```suggestion
# To pass custom parameters to the components, pass them as keyword arguments
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:385]:
```suggestion
# `async` code inside the methods, define them with `async def` and Ragna will
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:370]:
```suggestion
#     Unfortunately, Ragna's web UI currently does **not** support arbitrary custom
#     parameters. This is actively worked on and progress is tracked in
#     [#217](https://github.com/Quansight/ragna/issues/217). Until this is resolved,
#     the source storage and the assistant together take the following four
#     parameters:
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:377]:
```suggestion
#     When using Ragna's builtin components, all source storages cover parameters 1. to
```

File: docs/tutorials/gallery_custom_components.py
Comment (pavithraes) [inline docs/tutorials/gallery_custom_components.py:239]:
```suggestion
# ![Create-chat-modal of Ragna's web UI with the TutorialSourceStorage and TutorialAssistant as selected options](../../assets/images/ragna-tutorial-components.png)
```

**Review by pavithraes — CHANGES_REQUESTED** [review]
Thanks for working on this! Most of the suggestions are related to language being used, specifically [avoiding phrases like simply or just](https://developers.google.com/style/tone#some-things-to-avoid-where-possible) and using [active voice](https://developers.google.com/style/voice).

I have a note on my end to improve the existing tutorials to also use second person in the tutorials (i.e. "you" rather than "we") on this project, so not including those as review suggestions here.

File: docs/tutorials/gallery_custom_components.py
Comment (pmeier) [reply docs/tutorials/gallery_custom_components.py:152]: > let's suggest `sys.path` here?

That is not possible. `sys.path` is only available from inside Python, but when starting the API or UI, the user never runs Python directly. They have to make their custom components importable "from the outside". Let me try to rephrase this to make it more clear what needs to happen.

Comment (pmeier) [reply docs/tutorials/gallery_custom_components.py]: @pavithraes In 737e322e5039974f9c0cdb1973173b93e5f92833 I have shortened this section to only say that the code needs to be on the search path and included a link to our config reference where this is explained in more detail. Please have another look.

Comment (pavithraes) [reply docs/tutorials/gallery_custom_components.py]: Thanks, looks good :)

**Review by pavithraes — APPROVED** [review]
Thanks for the updates, looks good to me :)

**Pull request merged**
